README.md
# Node.js-LAN-File-Sharing
A small Node.js app designed for sharing files or chatting while on the same network. Especially useful when you are trying to get a file from a friend and their device has a single broken USB port or share clipboard.

## Features
- Easy to use Drag and Drop file upload.
- Faster than uploading to a server then downloading since you are the server.
- Works with large files (tested with >2gb).
- Fast loading and opening due to light weight (no library) Javascript and customized bootstrap.
- Node user account is not required for the chat

## How to use
0. Clone (or download as zip) this project from github
0. Open terminal (or command line) in project's folder.
0. First run ```npm install``` then ```node main.js```.
0. Project will display your LAN ip and port (8080 by default) in terminal once running.
0. On another device in the LAN, go to the url:port.
0. Drag & Drop any file in browser to transfer.

Project saves sent files in "files" folder. Also files in "files" folder can be downloaded from the browser.

PORT enviroment varible is used to change the port. e.g. ```export PORT=3030; node main.js```

## Screenshots

![Imgur](/screenshot/ss1.png)

Your LAN ip is shown in page in adition to console output. Files in "files" directory is also listed for download.  

![Imgur](/screenshot/ss1.png)

Drag and Drop file upload. Achived without any javascript library.

## TODO
I am planning to turn this project into a Electron app so people who are scared easly from terminals can use it as well.
Also pause/resume, progress bar and uploading folders are posible improvements I might implement in the future.


node-wins/CHANGELOG.md
# Node.js Changelog

Select a Node.js version below to view the changelog history:

* [Node.js 12](doc/changelogs/CHANGELOG_V12.md) - **Long Term Support**
* [Node.js 11](doc/changelogs/CHANGELOG_V11.md) - End-of-Life
* [Node.js 10](doc/changelogs/CHANGELOG_V10.md) — Long Term Support
* [Node.js 9](doc/changelogs/CHANGELOG_V9.md) — End-of-Life
* [Node.js 8](doc/changelogs/CHANGELOG_V8.md) — Long Term Support
* [Node.js 7](doc/changelogs/CHANGELOG_V7.md) — End-of-Life
* [Node.js 6](doc/changelogs/CHANGELOG_V6.md) — End-of-Life
* [Node.js 5](doc/changelogs/CHANGELOG_V5.md) — End-of-Life
* [Node.js 4](doc/changelogs/CHANGELOG_V4.md) — End-of-Life
* [io.js](doc/changelogs/CHANGELOG_IOJS.md) — End-of-Life
* [Node.js 0.12](doc/changelogs/CHANGELOG_V012.md) — End-of-Life
* [Node.js 0.10](doc/changelogs/CHANGELOG_V010.md) — End-of-Life
* [Archive](doc/changelogs/CHANGELOG_ARCHIVE.md)

Please use the following table to find the changelog for a specific Node.js
release.

<!--lint disable maximum-line-length-->
<table>
<tr>
  <th title="LTS Until 2022-04"><a href="doc/changelogs/CHANGELOG_V12.md">12</a><sup>LTS</sup></th>
  <th title="LTS Until 2021-04"><a href="doc/changelogs/CHANGELOG_V10.md">10</a><sup>LTS</sup></th>
  <th title="LTS Until 2019-12"><a href="doc/changelogs/CHANGELOG_V8.md">8</a><sup>LTS</sup></th>
</tr>
<tr>
    <td valign="top">
<b><a href="doc/changelogs/CHANGELOG_V12.md#12.13.1">12.13.1</a></b><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.13.0">12.13.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.12.0">12.12.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.11.1">12.11.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.11.0">12.11.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.10.0">12.10.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.9.1">12.9.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.9.0">12.9.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.8.1">12.8.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.8.0">12.8.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.7.0">12.7.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.6.0">12.6.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.5.0">12.5.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.4.0">12.4.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.3.1">12.3.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.3.0">12.3.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.2.0">12.2.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.1.0">12.1.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V12.md#12.0.0">12.0.0</a><br/>
    </td>
    <td valign="top">
<b><a href="doc/changelogs/CHANGELOG_V10.md#10.15.3">10.15.3</a></b><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.2">10.15.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.1">10.15.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.15.0">10.15.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.14.2">10.14.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.14.1">10.14.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.14.0">10.14.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.13.0">10.13.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.12.0">10.12.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.11.0">10.11.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.10.0">10.10.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.9.0">10.9.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.8.0">10.8.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.7.0">10.7.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.6.0">10.6.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.5.0">10.5.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.4.1">10.4.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.4.0">10.4.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.3.0">10.3.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.2.1">10.2.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.2.0">10.2.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.1.0">10.1.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V10.md#10.0.0">10.0.0</a><br/>
    </td>
    <td valign="top">
<b><a href="doc/changelogs/CHANGELOG_V8.md#8.16.0">8.16.0</a></b><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.15.1">8.15.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.15.0">8.15.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.14.1">8.14.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.14.0">8.14.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.13.0">8.13.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.12.0">8.12.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.11.4">8.11.4</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.11.3">8.11.3</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.11.2">8.11.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.11.1">8.11.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.11.0">8.11.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.10.0">8.10.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.9.4">8.9.4</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.9.3">8.9.3</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.9.2">8.9.2</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.9.1">8.9.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.9.0">8.9.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.8.1">8.8.1</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.8.0">8.8.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.7.0">8.7.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.6.0">8.6.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.5.0">8.5.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.4.0">8.4.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.3.0">8.3.0</a><br/>
<a href="doc/changelogs/CHANGELOG_V8.md#8.2.1">8.2.1</a><br/>

main.js
var open = require('open');
var http = require('http');
var fileShare = require('./app')({
    filesFolderPath: (process.argv[3] || null),
    port: (process.argv[2] || process.env.PORT),
    allowDeletion: false,
    progressCallback: function (progress, fileName) {
        //TODO: connect to UI when writing the electron app.
        console.log("Progress: " + fileName + " " + Math.floor(progress) + "%");
    },
    errorCallback: function (url, err) {
        if (err.status == 404) {
            console.log("(Not Found) " + url);
        } else {
            console.log("(errorCallback) " + url);
            console.error(err);
        }
    }
});

var server = http.createServer(fileShare.app);

// Init chat
require('./chat')(server);

function onError(error) {
    if (error.syscall !== 'listen') {
        throw error;
    }

    var bind = typeof port === 'string'
        ? 'Pipe ' + fileShare.port
        : 'Port ' + fileShare.port;

    // handle specific listen errors with friendly messages
    switch (error.code) {
        case 'EACCES':
            console.error(bind + ' requires elevated privileges');
            process.exit(1);
            break;
        case 'EADDRINUSE':
            console.error(bind + ' is already in use');
            process.exit(1);
            break;
        default:
            throw error;
    }
}

function onListening() {
    var addr = server.address();

    if (typeof addr === 'string') {
        console.log('Listening on pipe ' + addr);
    } else {
        fileShare.addresses.forEach(function (address) {
            var url = 'http://' + address + ':' + addr.port;
            console.log('Listening on ' + url);
            open(url);
        });
    }
}

server.listen(fileShare.port);
server.on('error', onError);
server.on('listening', onListening);

public/js/qr.js
var getServerInfo = function (callback) {
    var request = new XMLHttpRequest();
    request.open('GET', '/share/info', true);

    var onError = function (caller) {
        //TODO check if offline (or server is down)
        console.log(request);
    }

    request.onload = function () {
        if (request.status >= 200 && request.status < 400) {
            // Success!
            var resp = request.responseText;
            var data = JSON.parse(resp);
            if ("addresses" in data && "port" in data) {
                updateAddress(data["addresses"], data["port"]);
            }
        } else {
            onError(0);
        }
    };

    request.onerror = function () { onError(1); };

    request.send();
}

getServerInfo();
setInterval(function () {
    getServerInfo();
}, 2000);

public/js/chat.js
function copyText(e) {
    if (!window.isSecureContext) {
        return;
    }
    // var text = $(e).parent().parent().children('.messageBody').text()
    var text = $(e).children('.messageBody').text()
    navigator.clipboard.writeText(text).then(function () {
        /* clipboard successfully set */
        showSnackbar("Text has been copied to clipboard");
    }, function () {
        showSnackbar("Text cannot be copied");
        /* clipboard write failed */
    });
}

function showSnackbar(text = 'Unknown') {
    var div = document.getElementById("snackbar")
    div.innerHTML = text;
    div.className = "show";
    setTimeout(function () {
        div.className = div.className.replace("show", "");
    }, 3000);
}

$(function () {



    var FADE_TIME = 150; // ms
    var TYPING_TIMER_LENGTH = 400; // ms
    var COLORS = [
        '#e21400', '#91580f', '#f8a700', '#f78b00',
        '#58dc00', '#287b00', '#a8f07a', '#4ae8c4',
        '#3b88eb', '#3824aa', '#a700ff', '#d300e7'
    ];

    // Initialize variables
    var $window = $(window);
    var $usernameInput = $('.usernameInput'); // Input for username
    var $messages = $('.messages'); // Messages area
    var $inputMessage = $('.inputMessage'); // Input message input box

    var $loginPage = $('.login.page'); // The login page
    var $chatPage = $('.chat.page'); // The chatroom page

    // Prompt for setting a username
    var username;
    var connected = false;
    var typing = false;
    var lastTypingTime;
    var $currentInput = $usernameInput.focus();

    var socket = io();

    var addParticipantsMessage = function (data) {
        var message = '';
        if (data.numUsers === 1) {
            message += "there's 1 participant";
        } else {
            message += "there are " + data.numUsers + " participants";
        }
        log(message);
    }

    // Sets the client's username
    var setUsername = function () {
        username = cleanInput($usernameInput.val().trim());

        // If the username is valid
        if (username) {
            $loginPage.fadeOut();
            $chatPage.show();
            $loginPage.off('click');
            $currentInput = $inputMessage.focus();

            // Tell the server your username
            socket.emit('add user', username);
        }
    }

    // Sends a chat message
    var sendMessage = function () {
        var message = $inputMessage.val();
        // Prevent markup from being injected into the message
        message = cleanInput(message);
        // if there is a non-empty message and a socket connection
        if (message && connected) {
            $inputMessage.val('');
            addChatMessage({
                username: username,
                message: message
            });
            // tell server to execute 'new message' and send along one parameter
            socket.emit('new message', message);
        }
    }

    // Log a message
    var log = function (message, options) {
        var $el = $('<li>').addClass('log').text(message);
        addMessageElement($el, options);
    }

    // Adds the visual chat message to the message list
    var addChatMessage = function (data, options) {
        // Don't fade the message in if there is an 'X was typing'
        var $typingMessages = getTypingMessages(data);
        options = options || {};
        if ($typingMessages.length !== 0) {
            options.fade = false;
            $typingMessages.remove();
        }

        var $usernameDiv = $('<span class="username"/>')
            .text(data.username)
            .css('color', getUsernameColor(data.username));
        var $messageBodyDiv = $('<span class="messageBody">')
            .text(data.message);

        var $copy = $('<div class="badge badge-info">')
            .text('copy');

        var typingClass = data.typing ? 'typing' : '';
        var $messageDiv = $('<li class="message" onclick="copyText(this);"/>')
            .data('username', data.username)
            .addClass(typingClass)
            .append($usernameDiv, $messageBodyDiv);
        if (window.isSecureContext) {
            $messageDiv.append($copy);
        }

        addMessageElement($messageDiv, options);
    }

    // Adds the visual chat typing message
    var addChatTyping = function (data) {
        data.typing = true;
        data.message = 'is typing';
        addChatMessage(data);
    }

    // Removes the visual chat typing message
    var removeChatTyping = function (data) {
        getTypingMessages(data).fadeOut(function () {
            $(this).remove();
        });
    }

    // Adds a message element to the messages and scrolls to the bottom
    // el - The element to add as a message
    // options.fade - If the element should fade-in (default = true)
    // options.prepend - If the element should prepend
    //   all other messages (default = false)
    var addMessageElement = function (el, options) {
        var $el = $(el);

        // Setup default options
        if (!options) {
            options = {};
        }
        if (typeof options.fade === 'undefined') {
            options.fade = true;
        }
        if (typeof options.prepend === 'undefined') {
            options.prepend = false;
        }

        // Apply options
        if (options.fade) {
            $el.hide().fadeIn(FADE_TIME);
        }
        if (options.prepend) {
            $messages.prepend($el);
        } else {
            $messages.append($el);
        }
        $messages[0].scrollTop = $messages[0].scrollHeight;
    }

    // Prevents input from having injected markup
    var cleanInput = function (input) {
        return $('<div/>').text(input).html();
    }

    // Updates the typing event
    var updateTyping = function () {
        if (connected) {
            if (!typing) {
                typing = true;
                socket.emit('typing');
            }
            lastTypingTime = (new Date()).getTime();

            setTimeout(function () {
                var typingTimer = (new Date()).getTime();
                var timeDiff = typingTimer - lastTypingTime;
                if (timeDiff >= TYPING_TIMER_LENGTH && typing) {
                    socket.emit('stop typing');
                    typing = false;
                }
            }, TYPING_TIMER_LENGTH);
        }
    }

    // Gets the 'X is typing' messages of a user
    var getTypingMessages = function (data) {
        return $('.typing.message').filter(function (i) {
            return $(this).data('username') === data.username;
        });
    }

    // Gets the color of a username through our hash function
    var getUsernameColor = function (username) {
        // Compute hash code
        var hash = 7;
        for (var i = 0; i < username.length; i++) {
            hash = username.charCodeAt(i) + (hash << 5) - hash;
        }
        // Calculate color
        var index = Math.abs(hash % COLORS.length);
        return COLORS[index];
    }

    // Keyboard events

    $window.keydown(event => {
        // Auto-focus the current input when a key is typed
        if (!(event.ctrlKey || event.metaKey || event.altKey)) {
            $currentInput.focus();
        }
        // When the client hits ENTER on their keyboard
        if (event.which === 13) {
            if (username) {
                sendMessage();
                socket.emit('stop typing');
                typing = false;
            } else {
                setUsername();
            }
        }
    });

    $inputMessage.on('input', function () {
        updateTyping();
    });

    // Click events

    // Focus input when clicking anywhere on login page
    $loginPage.click(function () {
        $currentInput.focus();
    });

    // Focus input when clicking on the message input's border
    $inputMessage.click(function () {
        $inputMessage.focus();
    });

    // Socket events

    // Whenever the server emits 'login', log the login message
    socket.on('login', function (data) {
        connected = true;
        // Display the welcome message
        var message = "Welcome to Socket.IO Chat – ";
        log(message, {
            prepend: true
        });
        addParticipantsMessage(data);
    });

    // Whenever the server emits 'new message', update the chat body
    socket.on('new message', function (data) {
        addChatMessage(data);
    });

    // Whenever the server emits 'user joined', log it in the chat body
    socket.on('user joined', function (data) {
        log(data.username + ' joined');
        addParticipantsMessage(data);
    });

    // Whenever the server emits 'user left', log it in the chat body
    socket.on('user left', function (data) {
        log(data.username + ' left');
        addParticipantsMessage(data);
        removeChatTyping(data);
    });

    // Whenever the server emits 'typing', show the typing message
    socket.on('typing', function (data) {
        addChatTyping(data);
    });

    // Whenever the server emits 'stop typing', kill the typing message
    socket.on('stop typing', function (data) {
        removeChatTyping(data);
    });

    socket.on('disconnect', function () {
        log('you have been disconnected');
    });

    socket.on('reconnect', function () {
        log('you have been reconnected');
        if (username) {
            socket.emit('add user', username);
        }
    });

    socket.on('reconnect_error', function () {
        log('attempt to reconnect has failed');
    });

});

public/js/share.js
var previousFileNames = null;
var fileList = document.getElementById('fileList');
var updateFileList = function (fileNames, allowDeletion) {
    if (!(fileNames instanceof Array)) {
        fileList.innerHTML = "";
        return;
    }
    if (arrayEquals(previousFileNames, fileNames)) {
        return;
    }

    //TODO do this with document.createElement
    fileList.innerHTML = fileNames.map(function (fileName) {
        return `<div class="list-item">
        <a target="_blank" href="./f/`+ fileName + `" class="list-group-item" download>` + fileName + `</a>
        ${allowDeletion ? `<a class="list-group-item-close" href="javascript:void(0);" onclick="deleteFile('` + fileName + `')">&times;</a>` : ''}
      </div>`;
    }).join('');

    previousFileNames = fileNames;
}

function arrayEquals(ar1, ar2) {
    if (ar1 == null) {
        return ar2 == null;
    }
    if (ar2 == null) {
        return ar1 == null;
    }
    if (ar1.length != ar2.length) {
        return false;
    } else {
        for (var index = 0; index < ar1.length; index++) {
            if (ar1[index] != ar2[index]) return false;
        }
        return true;
    }
}

var deleteFile = function (fileName) {
    if (!confirm("Are you sure?")) return;
    var request = new XMLHttpRequest()
    request.open('GET', `./f/del/` + fileName, true)
    request.send();
}

var getServerInfo = function (callback) {
    var request = new XMLHttpRequest();
    request.open('GET', '/share/info', true);

    var onError = function (caller) {
        //TODO check if offline (or server is down)
        console.log(request);
    }

    request.onload = function () {
        if (request.status >= 200 && request.status < 400) {
            // Success!
            var resp = request.responseText;
            var data = JSON.parse(resp);
            if ("addresses" in data && "port" in data) {
                updateAddress(data["addresses"], data["port"]);
            }
            if ("fileList" in data) {
                updateFileList(data["fileList"], data["allowDeletion"]);
            }
        } else {
            onError(0);
        }
    };

    request.onerror = function () { onError(1); };

    request.send();
}

getServerInfo();
setInterval(function () {
    getServerInfo();
}, 2000);


/* Query based upload success error stuff */

var queryText = location.search;

//Clearing query.
window.history.pushState('obj', document.title, "http://" + location.host + location.pathname);

//TODO do this properly
var successIndex = queryText.startsWith("?success=");
if (successIndex) {
    var successPanel = document.getElementById('successPanel');
    if (successPanel) successPanel.className = successPanel.className.replace("hidden", "");

    var fileSuccessName = document.getElementById('fileSuccessName');
    fileSuccessName.innerText = decodeURIComponent(queryText.substring(9));

} else if (queryText.startsWith("?error=")) {
    var errorPanel = document.getElementById('errorPanel');
    if (errorPanel) errorPanel.className = errorPanel.className.replace("hidden", "");
}

/* Drag n Drop Upload */

var holder = document.body;
var fileToUploadButton = document.getElementById('fileToUploadButton');

holder.ondragover = function (e) {
    e.preventDefault();
    e.stopPropagation();
    var x = e.pageX;
    var y = e.pageY;

    fileToUploadButton.style.position = "fixed";
    fileToUploadButton.style.top = (y - 11) + "px";
    fileToUploadButton.style.left = (x - 40) + "px";

};

var fixButtonBack = function () { fileToUploadButton.style.position = "static"; }
holder.ondragend = fixButtonBack;
holder.ondragexit = fixButtonBack;
holder.ondragleave = function (e) {
    //console.log("holder.ondragleave",e);
    if (e.target == fileToUpload
        || e.pageX <= 10 || e.pageX >= window.innerWidth - 10
        || e.pageY <= 10 || e.pageY >= window.innerHeight - 10) {
        fixButtonBack();
    }
}
holder.ondrop = function (e) {
    if (e.target != fileToUpload) {
        e.preventDefault();
    }
    fixButtonBack();
}


var fileToUpload = document.getElementById('fileToUpload');
fileToUpload.onchange = function (e) {
    holder.className += " success";
    setTimeout(function () {
        fileToUpload.form.submit();
    }, 1);
}

app.js
var express = require('express');
var formidable = require('formidable');
var path = require('path');
var fs = require('fs');
var os = require('os');
var qr_image = require("qr-image");

/**
 * @param {string} basePath
 * @param {string} relativePath
 * @returns {Promise<Array<string>>}
 */
function recursiveReadDir(basePath) {
    return _readDirPromise(basePath).then((contents) => {
        return Promise.all(contents.map((fileOrDirName) => {
            let fileOrDirPath = path.join(basePath, fileOrDirName);
            return _isDirPromise(fileOrDirPath).then((isDir) => {
                if (isDir) {
                    return recursiveReadDir(fileOrDirPath);
                } else {
                    return [fileOrDirPath];
                }
            });
        }));
    })
        .then((arrayOfArrays) => {
            let result = [];
            for (let index = 0; index < arrayOfArrays.length; index++) {
                result.push(...arrayOfArrays[index]);
            }
            return result;
        })
}
function _readDirPromise(targetPath) {
    return new Promise((resolve) => {
        fs.readdir(targetPath, (err, contents) => {
            if (contents == null) {
                resolve([]);
            } else {
                resolve(contents.filter((fileName) => fileName[0] != '.'));
            }
        })
    });
}
function _isDirPromise(targetPath) {
    return new Promise((resolve) => {
        fs.lstat(targetPath, (err, stats) => {
            resolve(stats.isDirectory());
        });
    });
}

function normalizePort(val) {
    var port = parseInt(val, 10);

    if (isNaN(port)) {
        // named pipe
        return val;
    }

    if (port >= 0) {
        // port number
        return port;
    }

    return false;
}


module.exports = function (conf) {

    /*
        conf = {
            filesFolderPath:...,
            publicPath:...,
            port:...|8080,
            allowDeletion:false,
            progressCallback:...,
            errorCallback:...,
            progressThreshold:...|10,
            disable: {
                fileDownload:...|false,
                info:...|false
            }
        }
    */

    //Getting config from conf.
    var filesFolderPath = conf.filesFolderPath || path.join(__dirname, 'files'),
        publicPath = conf.publicPath || path.join(__dirname, 'public'),
        port = normalizePort(conf.port || '8080'),
        allowDeletion = conf.allowDeletion === true,
        progressCallback = conf.progressCallback || false,
        errorCallback = conf.errorCallback || false,
        progressThreshold = conf.progressThreshold || 10,
        disable = conf.disable || {};

    var interfaces = os.networkInterfaces();

    var addresses = [];
    for (var k in interfaces) {
        for (var k2 in interfaces[k]) {
            var address = interfaces[k][k2];
            // NOTE: Only handling IPv4 at the moment.
            if (address.family === 'IPv4' && !address.internal) {
                addresses.push(address.address);
            }
        }
    }

    let qrCodesPath = path.join(publicPath, "./qr_codes/");
    if (!fs.existsSync(qrCodesPath)) {
        fs.mkdirSync(qrCodesPath);
    }
    addresses.forEach((address) => {
        let qr_svg = qr_image.image(`http://${address}:${port}/`, { type: 'png' });
        qr_svg.pipe(fs.createWriteStream(path.join(publicPath, `./qr_codes/${address}_${port}.png`)));
    });

    //New express app
    var app = express();

    //For index. Basically app.get('/share/',...);
    app.use(express.static(publicPath));

    //For downloading files
    if (!disable.fileDownload) app.use('/share/f', express.static(filesFolderPath));

    app.get('/share/', function (req, res) {
        res.sendFile(path.join(publicPath, 'share.html'));
    });

    app.get('/chat/', function (req, res) {
        res.sendFile(path.join(publicPath, 'chat.html'));
    });

    app.get('/share/f/del/:filename', function (req, res) {

        if (allowDeletion) {
            var filename = req.params.filename

            try {
                fs.unlinkSync(`./files/` + filename)
                res.setHeader('Access-Control-Allow-Origin', '*');
                res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE'); // If needed
                res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,contenttype'); // If needed
                res.setHeader('Access-Control-Allow-Credentials', true); // If needed
                res.status(200)
                //file removed
            } catch (err) {
                err.status = 404;
                res.send(err);
            }
        } else {
            res.sendStatus(500);
        }

    });

    app.post('/share/', function (req, res) {

        // Bug fix for when the filesFolderPath folder does not exists
        if (!!conf.filesFolderPath) {
            // For a path that can have multiple non existent folders.
            // Borrowed from: https://stackoverflow.com/a/41970204
            filesFolderPath.split(path.sep).reduce((currentPath, folder) => {
                currentPath += folder + path.sep;
                if (!fs.existsSync(currentPath)) {
                    fs.mkdirSync(currentPath);
                }
                return currentPath;
            }, '');
        } else {
            // For the simple './files' path.
            if (!fs.existsSync(filesFolderPath)) {
                fs.mkdirSync(filesFolderPath);
            }
        }

        var form = new formidable.IncomingForm();

        form.parse(req);

        var finalName,
            progress;

        form.on('fileBegin', function (name, file) {

            progress = 0;

            fileName = file.name;
            var splitted = fileName.split(".");
            var extension, name;
            if (splitted.length > 1) {
                extension = splitted[splitted.length - 1];
                name = "";
                for (var i = 0; i < splitted.length - 1; i++) {
                    name += splitted[i];
                }
            } else {
                extension = "";
                name = fileName;
            }

            //For not overwriting files.
            var i = 0;
            while (fs.existsSync(path.join(filesFolderPath, fileName))) {
                fileName = name + " dup" + (i++) + "." + extension;
            }

            file.path = path.join(filesFolderPath, fileName);
            file.finalName = fileName;
            finalName = fileName;

        });

        form.on('file', function (name, file) {
            res.redirect('/share/?success=' + encodeURIComponent(file.finalName));
        });

        form.on('error', function (err) {
            res.redirect('/share/?error=1');
        });

        form.on('progress', function (bytesReceived, bytesExpected) {
            var temp = bytesReceived * 100 / bytesExpected;
            if (temp > progress + progressThreshold) {
                progress = Math.floor(temp);
                if (progressCallback) progressCallback(progress, finalName);
            }
        });

    });

    app.get('/share/info', function (req, res) {

        if (disable.info) {
            var err = new Error('Not Found');
            err.status = 404;
            res.send(err);
            return;
        }

        res.setHeader('Access-Control-Allow-Origin', '*');
        res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE'); // If needed
        res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,contenttype'); // If needed
        res.setHeader('Access-Control-Allow-Credentials', true); // If needed

        var info = { "addresses": addresses, "port": port, "allowDeletion": allowDeletion };

        if (disable.fileDownload) {
            res.json(info);
            return;
        }

        recursiveReadDir(filesFolderPath).then((foundPaths) => {
            info.fileList = foundPaths.map((foundPath) => {
                return path.relative(filesFolderPath, foundPath);
            })
            res.json(info);
        })

    });

    // catch 404
    app.use(function (req, res, next) {
        var err = new Error('Not Found');
        err.status = 404;
        next(err);
    });

    // development error handler
    app.use(function (err, req, res, next) {
        if (errorCallback) errorCallback(req.url, err);
        if (err.status == 404) {
            res.sendStatus(404);
        } else {
            res.sendStatus(500);
        }
    });

    app.set('port', port);

    return {
        "addresses": addresses,
        "app": app,
        "disable": disable, //For changing later.
        "port": port
    };

};
